text/0052-resource-importing-support.md

```diff
@@ -0,0 +1,272 @@
+# Support for importing existing resources into a CDK stack
+
+* **Original Author(s):**: @tomas-mazak
+* **Tracking Issue**: #52
+* **API Bar Raiser**: @rix0rrr
+
+Allow users to import resources, that were created manually or by different
+means, into CDK stacks.
+
+## Working Backwards - README
+
+```
+cdk import [STACK]
+
+Import existing resource(s) into the given STACK
+
+Options:
+[... standard CDK CLI and CDK CLI `deploy` options ...]
+  -o, --create-resource-mapping  If specified, CDK will generate a mapping of
+                                 existing physical resources to CDK constructs to
+                                 be imported as. The mapping will be written in
+                                 the given file path. No actual import operation
+                                 will be performed                      [string]
+  -m, --resource-mapping         If specified, CDK will use the given file to
+                                 map physical resources to CDK resources for
+                                 import, instead of interactively asking the
+                                 user. Can be run from scripts          [string]
+  -f, --force                    Allow importing resources with properties
+                                 different from the CDK construct configuration.
+                                 This will most likely lead to a stack drift.
+      --non-interactive          Suppress all interactive prompts. All
+                                 confirmation prompts assume yes, all resources
+                                 that can't be identified will be skipped from
+                                 import. (default: false if stdin is a TTY, true
+                                 otherwise)
+```
+
+Sometimes, it is useful to be able to import (enroll/adopt/...) AWS resources,
+that were created by different means (manually in AWS console, using different
+orchestrator, etc), into a CDK stack. Some resources can simply be deleted and
+recreated by CDK, but for others, this is not convenient: Typically stateful
+resources like S3 Buckets, DynamoDB tables etc, cannot be easily deleted without
+an impact on the service.
+
+To import an existing resource into a CDK stack:
+
+- Run a `cdk diff` to ensure there are no pending changes to the CDK stack you
```

Comment:
Is this something the CLI can identify automatically? If so, worth calling out.

(We should assume users don't read docs and will get into all kinds of bad situations if we'll let them.)

Reply:
Sounds like we're going to drop changes from the template before performing the import (from the text below). So that could/would include any unintended changes, and so this caveat doesn't really apply anymore?

Reply:
So there are two kinds of possible changes: either an already existing resources has its configuration changed and this is easily detectable and it's already handled by the PR code: https://github.com/aws/aws-cdk/pull/17666/files#diff-f0f2c0ae1f83424770eef7a1108e226ea791642f88a7cc18b8f41443b62479b9R120-R128

But for newly defined constructs/resources, I can't see how we could determine if user intent is to create them or import them. We interactively prompt user anyway so they will be able to skip the import (and we delete it from the template in such case for the import operation).

So yes, this sentence can be dropped. I was just thinking that it's a good practice to see if there's any difference between declared and deployed configuration, before the changes are made for the import operation. On the other hand, this is probably a good practice before _any_ changes to the code, not just import. So it might not belong here.

Reply:
dropped